### PR TITLE
Scc 2898

### PR DESCRIPTION
**Pull request by charmingduchess**
In kinesis_client.rb
- created retry_failed_records
- created decode_failed_records
- tests for those methods

### CODE REVIEW

File: lib/kinesis_client.rb [hunk 1]
```diff
@@ -99,8 +93,23 @@ def push_records
...
 
   def filter_failures(resp)
-    resp.records.filter_map.with_index do |record, i|
-      avro.decode(@records[i + @batch_size * @batch_count]) if record.responds_to?(:error_message)
+    failed_records = resp.records.filter_map.with_index do |record, i|
+      
+      @records[i + @batch_size * @batch_count] if record.responds_to?(:error_message)
+    end
+    @failed_records << failed_records
+    puts @failed_records
```
Comment (danamansana) [inline lib/kinesis_client.rb]: I don't think we want `puts` statements

File: lib/kinesis_client.rb [hunk 2]
```diff
@@ -99,8 +93,23 @@ def push_records
   end
 
   def filter_failures(resp)
-    resp.records.filter_map.with_index do |record, i|
-      avro.decode(@records[i + @batch_size * @batch_count]) if record.responds_to?(:error_message)
+    failed_records = resp.records.filter_map.with_index do |record, i|
+      
+      @records[i + @batch_size * @batch_count] if record.responds_to?(:error_message)
```
Comment (danamansana) [inline lib/kinesis_client.rb]: I'm a little nervous about `@batch_count` getting out of sync with the other variables. For example, when you modify `@records` in the `retry_failed_records` method, should `@batch_count` be updated?

Comment (mwbenowitz) [reply lib/kinesis_client.rb]: I think the way to ensure that this works as intended is to set `@batch_count = 0` at the start of `push_records`. I don't think it has any relevance outside of that context, right?

It even could just be a local variable there that gets passed in there, I don't think anything needs to interact with it from outside that method, it's just tracking our position within the current set of records. (or even just `each_slice.with_index` would be more foolproof?)

File: lib/kinesis_client.rb [hunk 3]
```diff
@@ -75,16 +77,8 @@ def push_batch(batch)
...
     })
-
-    if resp.failed_record_count > 0 
-      failure_message = {
-        failures: resp.failed_record_count,
-        failures_data: filter_failures(resp)
-      }
-      $logger.warn("Batch sent to #{config[:stream_name]} with failures: #{failure_message}")
-    else
-      $logger.info("Batch sent to #{config[:stream_name]} successfully")
```
Comment (mwbenowitz) [inline lib/kinesis_client.rb]: I'd like to keep these logging statements if possible! (or something close to them). Also agree with Daniel that anything that is being output with `puts` should be formally logged. (For many reasons, one of which being that we are looking at new instrumentation tools that can hypothetically pull monitoring data out of the structured logging output).

**Review by mwbenowitz — COMMENTED** [review]
This definitely gets the functionality we need, and I think it would work as-is, but as I said in my comments there are a few things that we can improve. Mostly the logging statements that you had before I thought were very useful and clear and I don't want to lose them!

**Comment by charmingduchess** [comment]
I removed the @batch_size instance variable and instead pass the batch that was just processed to the filter_failures method.

File: lib/kinesis_client.rb (hunk 3 above)
Comment (charmingduchess) [reply lib/kinesis_client.rb]: yea the puts was just leftover from debugging tests :+)

File: spec/kinesis_client/kinesis_client_spec.rb [hunk 4]
```diff
@@ -331,29 +278,52 @@ def error(*args)
...
-      allow(@mock_client).to receive(:put_records).and_return(@mock_failed_response)
-      @kinesis_client << '4'
-      @kinesis_client << '5'
-      @kinesis_client.push_records
-      @kinesis_client.decode_failed_records
-      expect(@kinesis_client.failed_records).to eql(['5'])
+  describe "#filter_failures" do
+    it "should return records and their error messages that do not enter the kinesis stream" do
+        @kinesis_client << '1'
+        @kinesis_client << '2'
```
Comment (mwbenowitz) [inline spec/kinesis_client/kinesis_client_spec.rb]: you shouldn't have to do this for this test. We should be only testing `filter_failures`. You should just be able to mock the batch in this test and then invoke the method. The rest of the test looks exactly right!

File: spec/kinesis_client/kinesis_client_spec.rb [hunk 5]
```diff
@@ -331,29 +278,52 @@ def error(*args)
...
+          {
+            data: "encoded 2",
+            partition_key: "hashed"
+          }
+        ],
+        stream_name: 'fake-stream'
+      }).and_return(@mock_failed_response)
+      @kinesis_client << '1'
+      @kinesis_client << '2'
+      @kinesis_client.push_records
```
Comment (mwbenowitz) [inline spec/kinesis_client/kinesis_client_spec.rb]: Same comment as above. We should be able to just set `@failed_records` and then mock `avro.decode` to output what we're testing.

**Review by mwbenowitz — COMMENTED** [review]
Just commenting on the tests: These work! but they can be simplified to not rely on invoking other methods in the module.

The general philosophy of unit testing is that a test should test one method and one method only. If we want to test the overall functionality of this module, we should add another layer of "functional" tests that test in more general (only mocking calls to external resources). But that is beyond the scope here.

**Review by danamansana — APPROVED** [review]

**Comment by charmingduchess** [comment]
Refactored #filter_failures and #failed_records tests to be more unit test like
- https://github.com/NYPL/NYPLRubyUtil/pull/14#pullrequestreview-866703098

**Pull request merged**